File: direct-mode/clients/ios/AvatarDemo/Components/Toast.swift
import SwiftUI

enum ToastKind {
    case error
    case warning

    var background: Color {
        switch self {
        case .error: return Color(red: 0.70, green: 0.15, blue: 0.12)
        case .warning: return Color(red: 0.54, green: 0.35, blue: 0.0)
        }
    }
}

struct ToastMessage: Equatable {
    let text: String
    var kind: ToastKind = .error
    /// Repeated identical failures must still re-trigger the animation, so the
    /// message carries its own identity rather than relying on text equality.
    let id = UUID()

    static func == (lhs: ToastMessage, rhs: ToastMessage) -> Bool { lhs.id == rhs.id }
}

/// Mirrors the web demos' Toast: SDK errors and blocked actions surface here
/// instead of only reaching a caption label.
struct ToastView: View {
    let message: ToastMessage

    var body: some View {
        Text(message.text)
            .font(.caption)
            .foregroundStyle(.white)
            .multilineTextAlignment(.center)
            .padding(.horizontal, 16)
            .padding(.vertical, 10)
            .background(message.kind.background)
            .cornerRadius(10)
            .padding(.horizontal, 16)
            .transition(.move(edge: .top).combined(with: .opacity))
    }
}

extension View {
    /// Presents `message` at the top of the screen and clears it after a beat.
    func toast(_ message: Binding<ToastMessage?>) -> some View {
        overlay(alignment: .top) {
            if let current = message.wrappedValue {
                ToastView(message: current)
                    .task(id: current.id) {
                        try? await Task.sleep(nanoseconds: 4_000_000_000)
                        withAnimation { message.wrappedValue = nil }
                    }
            }
        }
        .animation(.easeInOut(duration: 0.2), value: message.wrappedValue)
    }
}

File: direct-mode/clients/android/app/src/main/java/ai/spatius/avatarkit/directmodedemo/ui/Toast.kt
package ai.spatius.avatarkit.directmodedemo.ui

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay

enum class ToastKind { Error, Warning }

data class ToastMessage(
    val text: String,
    val kind: ToastKind = ToastKind.Error,
    // Repeated identical failures must still re-trigger the animation, so the
    // message carries a serial rather than relying on text equality.
    val serial: Long = 0L,
)

/**
 * Mirrors the web demos' Toast: SDK errors and blocked actions surface here
 * instead of only reaching the log panel.
 */
@Composable
fun ToastHost(
    message: ToastMessage?,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var visible by remember { mutableStateOf(false) }

    LaunchedEffect(message?.serial) {
        if (message == null) {
            visible = false
            return@LaunchedEffect
        }
        visible = true
        delay(4_000)
        visible = false
        // Let the exit animation finish before clearing the slot.
        delay(200)
        onDismiss()
    }

    Box(modifier = modifier.fillMaxWidth(), contentAlignment = Alignment.TopCenter) {
        AnimatedVisibility(
            visible = visible && message != null,
            enter = fadeIn() + slideInVertically { -it },
            exit = fadeOut() + slideOutVertically { -it },
        ) {
            val kind = message?.kind ?: ToastKind.Error
            val bg = when (kind) {
                ToastKind.Error -> Color(0xFFB3261E)
                ToastKind.Warning -> Color(0xFF8A5A00)
            }
            Text(
                text = message?.text.orEmpty(),
                color = Color.White,
                textAlign = TextAlign.Center,
                style = MaterialTheme.typography.bodySmall,
                modifier = Modifier
                    .padding(horizontal = 16.dp, vertical = 12.dp)
                    .clip(RoundedCornerShape(10.dp))
                    .background(bg)
                    .padding(horizontal = 16.dp, vertical = 10.dp),
            )
        }
    }
}
